README.md
# Reindexer

Little clojure project meant to be run on old servers for a few months, as a JAR compiled for java 6. Due to our old versions of rabbitmq and solr (and java!), I had to use very specific, and ancient, versions of java libraries and couldn't leverage more general-purpose clojure wrappers for RabbitMQ ([langohr](http://clojurerabbitmq.info/)) and solr (there's [flux](https://github.com/mwmitchell/flux) but it's... okay).

It expects JSON payloads on a very specific rabbitmq queue, and will add documents on a very specific Solr core (or set thereof).

It writes to a log file in its dir (`reindexer.log`), which uses log4j's [`RollingFileAppender`](https://logging.apache.org/log4j/1.2/apidocs/org/apache/log4j/RollingFileAppender.html) to keep its filesystem footprint small (keeps up to 20 1MB log files). For now it writes some INFO log messages when it's about to reindex, and it'll skip queue payloads that cause exceptions (dutifully logging their stacktraces, but not requeing). It's meant to chug along pretty much forever, unless an exception happens outside of the reindexing function--in which case it'll neatly close its connection to RMQ before dying.

For example:

If you manually put in this message (which is a full product representation):

```json
[{"slug":"blow-up-thickening-mist","is_men":false,"is_women":true,"is_searchable":true,"updated_at":"2018-10-22T02:11:34Z","id":"111","magento_id":"111","url":"/shop/blow-up-thickening-mist","title":"HELLO FROM UBERJAR","product_name":"blowPro Blow Up Thickening Mist","sku":"BLOW-MIS-THICK-FZ","type":"simple","family":"default","product_notes":"","ship_with_box":0,"presale_only":0,"mes_title":"","mes_short_description":"","mes_full_description":"HERES THE CHANGES","mes_size":"","mes_additional_notes":"","is_changeable_subscription":0,"is_lifestyle":null,"mes_group_id":"","key_ingredients":"","great_for":"YOUR HAIR","meta_title":"Blow Up Thickening Mist ","meta_keyword":"","meta_description":"","images":"{\"pdp\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/384x384/163b81649b7ef7bc8a00b0066e59ae0a/t/e/test-image-2-1.jpg\",\"width\":384,\"height\":384,\"is_placeholder\":false},\"box_list\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/200x176/163b81649b7ef7bc8a00b0066e59ae0a/t/e/test-image-2-1.jpg\",\"width\":200,\"height\":176,\"is_placeholder\":false},\"box_feature\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/640x528/163b81649b7ef7bc8a00b0066e59ae0a/t/e/test-image-2-1.jpg\",\"width\":640,\"height\":528,\"is_placeholder\":false}}","media_s":"[{\"format\":\"image\",\"type\":\"primary\",\"position\":1,\"sizes\":{\"large\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/384x384/163b81649b7ef7bc8a00b0066e59ae0a/t/e/test-image-2-1.jpg\",\"width\":384,\"height\":384,\"is_placeholder\":false},\"small\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/96x96/163b81649b7ef7bc8a00b0066e59ae0a/t/e/test-image-2-1.jpg\",\"width\":96,\"height\":96,\"is_placeholder\":false},\"xlarge\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/small_image/1500x1700/b38cf51ec77170b109c5e310157197eb/t/e/test-image-2-1.jpg\",\"width\":1500,\"height\":1700,\"is_placeholder\":false}}},{\"format\":\"image\",\"type\":\"swatch\",\"position\":5,\"sizes\":{\"large\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/thumbnail/384x384/163b81649b7ef7bc8a00b0066e59ae0a/b/l/blow_blowup_900x900.jpg\",\"width\":384,\"height\":384,\"is_placeholder\":false},\"small\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/thumbnail/96x96/163b81649b7ef7bc8a00b0066e59ae0a/b/l/blow_blowup_900x900.jpg\",\"width\":96,\"height\":96,\"is_placeholder\":false},\"xlarge\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/thumbnail/900x900/b38cf51ec77170b109c5e310157197eb/b/l/blow_blowup_900x900.jpg\",\"width\":900,\"height\":900,\"is_placeholder\":false}}},{\"format\":\"image\",\"type\":\"sample\",\"sizes\":{\"large\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/sample_image/384x384/163b81649b7ef7bc8a00b0066e59ae0a/images/catalog/product/placeholder/sample_image.jpg\",\"width\":384,\"height\":384,\"is_placeholder\":true},\"small\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/sample_image/96x96/163b81649b7ef7bc8a00b0066e59ae0a/images/catalog/product/placeholder/sample_image.jpg\",\"width\":96,\"height\":96,\"is_placeholder\":true},\"xlarge\":{\"url\":\"https://www.luis-beta.dev.birchbox.com/shop/media/catalog/product/cache/1/sample_image/b38cf51ec77170b109c5e310157197eb/images/catalog/product/placeholder/sample_image.jpg\",\"width\":262,\"height\":262,\"is_placeholder\":true}}}]","ingredients":[],"ingredients_md":"","brand":"blowPro","brand_id":97,"description":"Flat, limp hair? New York\u2019s blow dry experts have you covered. blowPro's thickening mist is the ultimate way to pump up hair without weighing it down \u2014 perfect for daily use and a must for voomy up-dos.","description_md":"\u003Cp\u003EFlat, limp hair? New York\u2019s blow dry experts have you covered. blowPro's thickening mist is the ultimate way to pump up hair without weighing it down \u2014 perfect for daily use and a must for voomy up-dos.\u003C/p\u003E","short_description":"Give ho-hum hair a boost with this lightweight thickening mist. ","expert_take":"Revive flat hair with blowPro Faux Dry Shampoo.","expert_take_md":"\u003Cp\u003ERevive flat hair with \u003Ca href=\"/shop/blow-faux-dry-shampoo\"\u003EblowPro Faux Dry Shampoo\u003C/a\u003E.\u003C/p\u003E","how_it_works":"","how_to_use":"Spritz lightly on clean, damp hair in sections and blow-dry in upward motions for all over lift.","how_to_use_md":"\u003Cp\u003ESpritz lightly on clean, damp hair in sections and blow-dry in upward motions for all over lift.\u003C/p\u003E\r\n","details_html":"\u003Ch2\u003EBirchbox Breakdown\u003C/h2\u003E\n    \u003Cp\u003EFlat, limp hair? New York\u2019s blow dry experts have you covered. blowPro's thickening mist is the ultimate way to pump up hair without weighing it down \u2014 perfect for daily use and a must for voomy up-dos.\u003C/p\u003E\n    \u003Ch2\u003EHow To Use\u003C/h2\u003E\n    \u003Cp\u003ESpritz lightly on clean, damp hair in sections and blow-dry in upward motions for all over lift.\u003C/p\u003E\n    \u003Ch2\u003EEditor's Tip\u003C/h2\u003E\n    \u003Cp\u003ERevive flat hair with \u003Ca href=\"birchbox://view/products/107\"\u003EblowPro Faux Dry Shampoo\u003C/a\u003E.\u003C/p\u003E\n","size":"8.5 oz","brand_ids":[97],"related_product_ids":[],"is_discontinued":false,"rating":"3.4","rating_weighted":"4.2615","rating_purchased":"3.3333","rating_sampled":"0.0","weighted_rating_sf":"4.2615","max_sale_qty":50,"review_count":14,"in_stock":true,"is_free_shipping":false,"is_reviewable":true,"is_favoritable":true,"is_exclusive":false,"is_visible_in_catalog":true,"is_hazmat":true,"price":"21.0","currency":"USD","blurb":"","upc":"","weight":"1.0","availability":"available","swatch_hex_code":"","category_names":["Hair","Birchbox: December 2010","Birchbox: January 2011 Welcome","Birchbox: March 2011","March's Box - 2011","$25 and Under","Styling Products","Protective Sprays","Texture \u0026 Beach Sprays"],"category_position_6":24,"category_position_101":1,"category_position_157":0,"category_position_194":1,"category_position_196":1,"category_position_217":1,"category_position_580":25,"category_position_589":20,"category_position_697":16,"category_position_3372":13,"category_position_3386":12,"brand_position_97":0,"category_urls":["/","//shop","//shop/march-box","//shop/hair","//shop/hair/treatments","//shop/hair/treatments/protective-sprays","//shop/hair/styling-products-1","//shop/hair/styling-products-1/texture-beach-sprays","//shop/featured","//shop/featured/25-and-under-1","//shop/featured/25-and-under-1/hair","//shop/birchbox-1","//shop/birchbox-1/march-2011","//shop/birchbox-1/march-2011/birchbox-march-2011-bb6","//shop/birchbox-1/march-2011/birchbox-march-2011-bb4","//shop/birchbox-1/january-2011","//shop/birchbox-1/january-2011/birchbox-january-2011-welcomeb","//shop/birchbox-1/apr-dec-2010","//shop/birchbox-1/apr-dec-2010/birchbox-december-2010-volume"],"category_ids":["2","8","280","101","277","196","194","275","157","6","698","3372","697","3386","217","15","580","589"]}]

```

(Ideally an API puts the messages, but let's say it's from RabbitMQ's admin interface):

![image](https://user-images.githubusercontent.com/82133/47277035-e5f3dc00-d589-11e8-99b6-40bb01d3683f.png)


It'll consume it and chuck it onto solr:

![image](https://user-images.githubusercontent.com/82133/47277021-b80e9780-d589-11e8-966f-3b3459135a04.png)

If you tail its log, you may see this (notice that, since `map` is lazy, this first message will open the connections to the solr cores):

```
[2018-10-23 01:06:06,111][INFO][reindexer.core] About to index: 111
[2018-10-23 01:06:06,120][INFO][org.apache.solr.client.solrj.impl.HttpClientUtil] Creating new http client, config:maxConnections=128&maxConnectionsPerHost=32&followRedirects=false
```

And if you send a bad payload, it'll log that error (in this case, malformed JSON) and keep chugging along:

```
[2018-10-23 01:05:30,376][ERROR][reindexer.core] Error reindexing, skipping!
java.lang.Exception: JSON error (expected false)
	at clojure.data.json$_read.invokeStatic(json.clj:215)
	at clojure.data.json$_read.invoke(json.clj:177)
	at clojure.data.json$read.invokeStatic(json.clj:272)
	at clojure.data.json$read.doInvoke(json.clj:228)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:648)
	at clojure.core$apply.invoke(core.clj:641)
	at clojure.data.json$read_str.invokeStatic(json.clj:278)
	at clojure.data.json$read_str.doInvoke(json.clj:274)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at reindexer.core$json_reindexer$fn__409.invoke(core.clj:26)
	at reindexer.rabbitmq$consume_delivery.invokeStatic(rabbitmq.clj:47)
	at reindexer.rabbitmq$consume_delivery.invoke(rabbitmq.clj:42)
	at reindexer.rabbitmq$consume_forever.invokeStatic(rabbitmq.clj:55)
	at reindexer.rabbitmq$consume_forever.invoke(rabbitmq.clj:50)
	at reindexer.rabbitmq$subscribe_to_queue.invokeStatic(rabbitmq.clj:68)
	at reindexer.rabbitmq$subscribe_to_queue.invoke(rabbitmq.clj:62)
	at clojure.core$partial$fn__4759.invoke(core.clj:2515)
	at reindexer.core$_main.invokeStatic(core.clj:40)
	at reindexer.core$_main.doInvoke(core.clj:34)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at reindexer.core.main(Unknown Source)

```


## Usage

Ideally, you should grab the standalone uberjar and chuck it on a server:

    $ java -jar uberjar/reindexer-0.1.0-SNAPSHOT-standalone.jar

However, it needs some env vars; see the example shell script for the env variables that it expects to pick up, so you can use `run.sh.example` as a basis for something like:

    $ ./reindexer.sh


## Development

See the `comment` blocks at the end of each of the source files for some thoughts on interactive development (inspired by Stuart Halloway's ["running with scissors" talk](https://www.youtube.com/watch?v=Qx0-pViyIDU))

I'm using `leiningen` for project management, Emacs + CIDER for interactive development, and a couple of ssh tunnels to my development ec2 instance to leverage its rabbitmq and solr (need to have an ssh session for these to be picked up, natch):

```

Host new-dev
  Hostname luis.dev.birchbox.com
  ForwardAgent yes
  User luis
  LocalForward 5673 rabbitmq:5672
  LocalForward 8081 solr:8080
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-10-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-10-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/reindexer/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/reindexer/compare/0.1.0...0.1.1


test/reindexer/core_test.clj
(ns reindexer.core-test
  (:require [clojure.test :refer :all]
            [reindexer.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject reindexer "0.1.0-SNAPSHOT"
  :description "Lil reindexing daemon"
  :url ""
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/data.json "0.2.6"]
                 [org.clojure/tools.logging "0.4.1"]
                 [org.slf4j/slf4j-log4j12 "1.6.2"]
                 [log4j "1.2.16"]
                 ; these older versions match our sad prod/dev java/solr
                 [com.rabbitmq/amqp-client "4.8.3"]
                  ; this bullshit is required by the http solr server:
                 [commons-logging/commons-logging "1.1.1"]
                 [org.apache.solr/solr-solrj "4.4.0"]]
  :main ^:skip-aot reindexer.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/reindexer/solr.clj
(ns reindexer.solr
  (:import [org.apache.solr.client.solrj SolrQuery]
           [org.apache.solr.client.solrj.impl HttpSolrServer]
           [org.apache.solr.common SolrInputDocument]))

(defn connect-to-cores
  "Given a config map, connects to a remote solr instance via http"
  [urls core-name]
  (let [core-urls (map #(str % "/" (name core-name)) urls)]
    (map #(HttpSolrServer. %) core-urls)))

(defn query-solr
  "Given a solr cxn and a query, return a collection of results"
  [client query-string]
  (let [query    (doto (SolrQuery.) (.setQuery query-string))
        response (.query client query)]
    (.getResults response)))

(defn as-input-document
  "Given a clojure map with string keys, return a solr input document"
  [m]
  (let [document (SolrInputDocument.)
        ; filter out nils, empty strings and empty arrays
        allowed-value? #(not (or (nil? %)
                                 (some #{"" []} #{%})))]
    (doseq [[name value] m
            :when (allowed-value? value)]
      (.addField document name value))
    document))

(defn update-in-cores
  "Given an array of connections and document maps, update documents in the colls"
  [clients document-maps]
  (let [documents (map as-input-document document-maps)]
    (doseq [c clients]
      (.add c documents)
      (.commit c))))

(defn remove-from-cores
  "Given an array of connections and ids, remove the identified documents"
  [clients id-list]
  (doseq [c clients]
    (.deleteById id-list)
    (.commit c)))

;; Some notes
;; Had to tread carefully because solrj ifaces are always changing, thankfully
;; the javadocs + sorta current guides kinda helped:
;; * https://lucene.apache.org/solr/5_5_0//solr-solrj/org/apache/solr/client/solrj/package-summary.html
;; * https://lucene.apache.org/solr/guide/6_6/using-solrj.html
;; * http://www.solrtutorial.com/solrj-tutorial.html

; Have to add a localtunnel to my ssh config for a remote box:
;; LocalForward 8081 solr:8080

;; some REPL testing: I usually send forms one by one with C-c M-p
;; but you can also just eval the whole thing with C-c M-e at the end of the
;; do block.

(comment
  (do
    (def clients (connect-to-cores ["http://127.0.0.1:8081/solr440/products"]))
    (def results (query-solr (first clients) "id:111"))
    (println (get (first results) "product_name"))
                                        ; => "blowPro Blow Up Thickening Mist
    (def doc (first results))
    ;; the document is a SolrDocument instance, so, mutable
    (.setField doc "product_name" "Luis's magical product")
    (.get doc "product_name")
    (def doc-map (zipmap (keys doc) (vals doc)))
    ; notice it's a clojure map now:
    (get doc-map "product_name")
    ;; one interesting to notice is that trying to print new-doc
    ;; to the REPL raises an exception. Gotta inspect it lightly
    (def new-doc (as-input-document doc-map))
    (.keySet new-doc)
    (.get new-doc "product_name")
    ;; notice that "title_sort" isn't supposed to be here, but that's a testing
    ;; issue, so we just dissoc:
    (def doc-maps [(dissoc doc-map "title_sort")])
    (update-in-cores clients doc-maps)))


src/reindexer/rabbitmq.clj
(ns reindexer.rabbitmq
  (:import [com.rabbitmq.client
            Connection
            Channel
            ConnectionFactory
            QueueingConsumer]))

;; core methods (see usage at end of file)

(defn connect
  "Takes some config, returns a connection to RMQ"
  [{:keys [username password virtual-host host port]
    :as config}]
  (-> (doto (ConnectionFactory.)
        (.setUsername username)
        (.setPassword password)
        (.setVirtualHost virtual-host)
        (.setHost host)
        (.setPort port))
      (.newConnection)))

(defn setup-channel
  "Takes a connection and config and sets up a channel with:
  * a durable, non-autodelete exchange of 'direct' type
  * a durable, non-exclusive, non-autodelete queue with a well-known name
  Returns the channel"
  [cxn {:keys [queue-name exchange-name routing-key]
        :as config}]
  (let [channel (.createChannel cxn)]
    (.exchangeDeclare channel exchange-name "direct" true)
    (.queueDeclare channel queue-name true false false nil)
    (.queueBind channel queue-name exchange-name routing-key)
    channel))

(defn create-consumer
  "Takes a channel and queue name and creates a QueuingConsumer for it"
  [channel queue-name]
  (let [consumer (QueueingConsumer. channel)]
    (.basicConsume channel queue-name false consumer)
    consumer))

(defn consume-delivery
  "Takes a channel and a delivery, applies a processor to the msg; acks"
  [channel delivery processor]
  (let [delivery-tag  (-> delivery .getEnvelope .getDeliveryTag)
        delivery-body (slurp (.getBody delivery))]
    (processor delivery-body)
    (.basicAck channel delivery-tag false)))

(defn consume-forever
  "Takes a channel/cxn, a consumer and a processor; consumes forever.
  Dies upon exception, closing the channel and connection"
  [connection channel consumer processor]
  (try (loop []
         (consume-delivery channel
                           (.nextDelivery consumer)
                           processor)
         (recur))
       (finally (.close channel)
                (.close connection))))

(defn subscribe-to-queue
  "Takes some config and a processor, sets up stuff, consumes forever"
  [config processor]
  (let [connection (connect config)
        channel    (setup-channel connection config)
        consumer   (create-consumer channel (:queue-name config))]
    (consume-forever connection channel consumer processor)))

;; Some notes:

;; Followed this guide:
;; https://www.rabbitmq.com/api-guide.html#connecting
;; Notice that I had to set my local tunnel as follows:
;; Host new-dev
;;   Hostname luis-beta.dev.birchbox.com
;;   ForwardAgent yes
;;   User luis
;;   LocalForward 5673 rabbitmq:5672
;; As per the "service catalog": https"//birchbox.atlassian.net/wiki/spaces/TECH/pages/43155572/Service+Catalog#ServiceCatalog-examplesExamples"
;; was able to see my connections and stuff here: http://rabbitmq.luis-beta.dev.birchbox.com:15672/#/connections
;; There's some examples of producers and consumers in rubyland here:
;; https://github.com/birchbox/authentication_sdk/blob/ddac058ab6e8c5413cf58787f60dc724e74b3a53/lib/authentication_sdk/publisher.rb#L13-L16
;; https://github.com/birchbox/percolator/blob/master/lib/consumer/base_consumer.rb
;; and walked through here: https://birchbox.atlassian.net/wiki/spaces/TECH/pages/62586906/Lifecycle+user+API+session+to+asset+broker+visitor+id+entry

;; reindexer.rabbitmq> (def resp (.basicGet chan "amq.gen-hl_RMM1OkUd7q7iNLDob-w" true))
;; #'reindexer.rabbitmq/resp
;; reindexer.rabbitmq> (slurp (.getBody resp))
;; "{\"hello\":\"world\"}"
;; reindexer.rabbitmq> 

;; followed also: 
;; http://www.learningclojure.com/2011/02/rabbitmq-clojure-hello-world.html
;; https://www.rabbitmq.com/releases/rabbitmq-java-client/v2.1.1/rabbitmq-java-client-javadoc-2.1.1/com/rabbitmq/client/QueueingConsumer.html

 ;; // Create connection and channel.
 ;; ConnectionFactory factory = new ConnectionFactory();
 ;; Connection conn = factory.newConnection();
 ;; Channel ch1 = conn.createChannel();

 ;; // Declare a queue and bind it to an exchange.
 ;; String queueName = ch1.queueDeclare().getQueue();
 ;; ch1.queueBind(queueName, exchangeName, queueName);

 ;; // Create the QueueingConsumer and have it consume from the queue
 ;; QueueingConsumer consumer = new QueueingConsumer(ch1);
 ;; ch1.basicConsume(queueName, false, consumer);

 ;; // Process deliveries
 ;; while (/* some condition * /) {
 ;;     QueueingConsumer.Delivery delivery = consumer.nextDelivery();
 ;;     // process delivery
 ;;     ch1.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
;; }

;; More references for RMQ:

;; * https://www.rabbitmq.com/api-guide.html#connecting
;; * https://www.rabbitmq.com/releases/rabbitmq-java-client/v2.1.1/rabbitmq-java-client-javadoc-2.1.1/com/rabbitmq/client/QueueingConsumer.html
;; * https://www.cloudamqp.com/blog/2015-09-03-part4-rabbitmq-for-beginners-exchanges-routing-keys-bindings.html

;; example usage:
;; (you can use C-c M-p to throw sexps from here into the REPL)
;; as per: https://cider.readthedocs.io/en/latest/interactive_programming/

(comment
  (do (def cxn (connect))
      (def example-channel (setup-channel cxn
                                          {:queue-name "reindex"
                                           :exchange-name "product-reindexer"
                                           :routing-key "us.product"}))
      (def example-consumer (create-consumer example-channel "reindex"))
      (println "At this point, you'd use some tool to publish a message")
      (consume-delivery example-channel
                        (.nextDelivery example-consumer)
                        println)
      (println "The above call would block until there's a delivery,
                    and return hello world eventually")
      (println "Or you can e.g. consume up to 3 messages:")
      (let [cnt (atom 3)]
        (while (pos? @cnt)
          (do
            (consume-delivery example-channel
                              (.nextDelivery example-consumer)
                              #(println (str "received: " %)))
            (swap! cnt dec))))
      (println "Also cool to just consumer forever:")
      (consume-forever example-channel
                       example-consumer
                       #(println (str "got: " %)))
      ;; e.g.
      ;; => got: hello world 1
      ;; => got: hello world 2
      ;; ...
      (println "You probably killed the above; it closed the channel")
      (println "Let's try with some toy config:")
      (def config {:username "luis"
                   :password "hunter2"
                   :virtual-host "/birchbox-event-bus"
                   :host "127.0.0.1"
                   :port 5673
                   :queue-name "reindex"
                   :exchange-name "reindex-events"
                   :routing-key "product_reindex"})
      (defn print-message [message]
        (println (str "Received: " message)))
      (subscribe-to-queue config print-message)
      (.close cxn)))


src/reindexer/core.clj
(ns reindexer.core
  (:gen-class)
  (:require [reindexer.rabbitmq :as rmq]
            [reindexer.solr :as solr]
            [clojure.data.json :as json]
            [clojure.tools.logging :as log]))

(defn build-config
  "Grabs config from the environment"
  []
  {:username (System/getenv "RMQ_USERNAME")
   :password (System/getenv "RMQ_PASSWORD")
   :virtual-host (System/getenv "RMQ_VHOST")
   :host (System/getenv "RMQ_HOST")
   :port (Integer/parseInt (System/getenv "RMQ_PORT"))
   :queue-name (System/getenv "RMQ_QUEUE_NAME")
   :exchange-name (System/getenv "RMQ_EXCHANGE_NAME")
   :routing-key (System/getenv "RMQ_ROUTING_KEY")
   :solr-urls (clojure.string/split (System/getenv "SOLR_URLS") #",")})

(defn json-reindexer
  "Takes a config and returns a closed-over fn that can update solr cores from JSON"
  [config]
  (let [clients (solr/connect-to-cores (:solr-urls config))]
    (fn [message-body]
      (try (let [json-maps (json/read-str message-body)]
             (log/info (str "About to index: "
                            (clojure.string/join ","
                                                 (map #(get % "id") json-maps))))
             (solr/update-in-cores clients json-maps))
           (catch Exception e
             (log/error e "Error reindexing, skipping!"))))))

(defn -main
  "Infinite loop that picks up reindex messages and sends them to solr"
  [& args]
  (let [config       (build-config)
        sub-forever  (partial rmq/subscribe-to-queue config)
        processor    (json-reindexer config)]
    (sub-forever processor)))

;; ======= DEV NOTES FOLLOW =================================

;; I chuck these s-exps from here to a running CIDER repl
;; with C-c M-p, useful to put the whole let in there and have the
;; "process" running on a terminal
(comment
  ; Ideally, all that -main does is pick up stuff from the env
  ; and then spins up two closures from the libs and lets
  ; them run forever
  (defn json-consumer
    "Helps with repl debugging"
    [config]
    (fn [message-body]
      (clojure.pprint/pprint (json/read-str message-body))))
  (defn json-exhauster [_] #(println %))
  (let [config       {:username "luis"
                      :password "hunter2"
                      :virtual-host "/birchbox-event-bus"
                      :host "127.0.0.1"
                      :port 5673
                      :queue-name "reindex"
                      :exchange-name "reindex-events"
                      :routing-key "product_reindex"
                      :solr-urls (clojure.string/split "http://127.0.0.1:8081/solr440/products" #",")}
        sub-forever  (partial rmq/subscribe-to-queue config)
        ; can also send json-consumer if you're playing around in
        ; the REPL
        processor    (json-reindexer config)]
    (sub-forever processor)))

;; NOTES

; Had to do all this bullshit to produce some sort of workable JSON from rails:
;; coll = ProductCollection.new(ids: [111], indexable: true)
;; products = Solr::ProductCollectionDecorator.decorate(coll, preload_data: false, discontinued_only: false, context: {bust_caches: true})
;; batch_products = Solr::ProductCollectionDecorator.new(coll, products: products.products)
;; h = Solr::ProductCollectionSerializer.new(batch_products, root: false).serializable_hash[:products]
;; puts h.to_json # then copy-paste this verbatim in the rabbitmq env
;; h.as_json.first["brand_position_97"]
;; however, that last datum returned "" instead of 0,
;; causing the following error in the remote solr:
;; (from http://solr.luis-beta.dev.birchbox.com:8080/solr440/#/~logging)
;; null:org.apache.solr.common.SolrException: Error while creating field 'brand_position_97{type=sint,properties=indexed,stored,omitNorms,sortMissingLast}' from value ''

